tests/terraform/manual_upgrade/upgradecluster_test.go

```diff
@@ -0,0 +1,396 @@
+	flag.Parse()
+
+	RunSpecs(t, "Upgrade Cluster Test Suite")
+}
+
+var _ = Describe("Upgrade Tests:", func() {
+	Context("Build Cluster:", func() {
+		It("Starts up with no issues", func() {
+			status, err := terraform.BuildCluster(&testing.T{}, *tfVars, false)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(status).To(Equal("cluster created"))
+			defer GinkgoRecover()
+			fmt.Println("Server Node IPS:", terraform.MasterIPs)
+			fmt.Println("Agent Node IPS:", terraform.WorkerIPs)
+			terraform.PrintFileContents(terraform.KubeConfigFile)
+			Expect(terraform.MasterIPs).ShouldNot(BeEmpty())
+			if terraform.NumWorkers > 0 {
+				Expect(terraform.WorkerIPs).ShouldNot(BeEmpty())
+			} else {
+				Expect(terraform.WorkerIPs).Should(BeEmpty())
+			}
+			Expect(terraform.KubeConfigFile).ShouldNot(BeEmpty())
+		})
+
+		It("Checks Node and Pod Status", func() {
+			defer func() {
+				_, err := terraform.Nodes(terraform.KubeConfigFile, true)
+				if err != nil {
+					fmt.Println("Error retrieving nodes preupgrade: ", err)
+				}
+				_, err = terraform.Pods(terraform.KubeConfigFile, true)
+				if err != nil {
+					fmt.Println("Error retrieving pods preupgrade: ", err)
+				}
+			}()
+
+			fmt.Printf("\nFetching node status preupgrade\n")
+			expectedNodeCount := terraform.NumServers + terraform.NumWorkers
+			Eventually(func(g Gomega) {
+				nodes, err := terraform.Nodes(terraform.KubeConfigFile, false)
+				g.Expect(err).NotTo(HaveOccurred())
+				g.Expect(len(nodes)).To(Equal(expectedNodeCount), "Number of nodes should match the spec")
+				for _, node := range nodes {
+					g.Expect(node.Status).Should(Equal("Ready"), "Nodes should all be in Ready state")
+				}
+			}, "420s", "5s").Should(Succeed())
+
+			re := regexp.MustCompile("[0-9]+")
+			fmt.Printf("\nFetching pod status preupgrade\n")
+			Eventually(func(g Gomega) {
+				pods, err := terraform.Pods(terraform.KubeConfigFile, false)
+				g.Expect(err).NotTo(HaveOccurred())
+				for _, pod := range pods {
+					if strings.Contains(pod.Name, "helm-install") {
+						g.Expect(pod.Status).Should(Equal("Completed"), pod.Name)
+					} else {
+						g.Expect(pod.Status).Should(Equal("Running"), pod.Name)
+						g.Expect(pod.Restarts).Should(Equal("0"), pod.Name)
+						numRunning := re.FindAllString(pod.Ready, 2)
+						g.Expect(numRunning[0]).Should(Equal(numRunning[1]), pod.Name, "should have all containers running")
+					}
+				}
+			}, "600s", "5s").Should(Succeed())
+		})
+	})
+
+	Context("Preupgrade Validations:", func() {
+		It("Verifies ClusterIP Service Preupgrade", func() {
+			namespace := "auto-clusterip"
+			_, err := terraform.DeployWorkload("clusterip.yaml", terraform.KubeConfigFile)
+			Expect(err).NotTo(HaveOccurred(), "Cluster IP manifest not deployed")
+
+			Eventually(func(g Gomega) {
+				cmd := "kubectl get pods -n " + namespace + " -o=name -l k8s-app=nginx-app-clusterip --field-selector=status.phase=Running --kubeconfig=" + terraform.KubeConfigFile
```

Comment:
There multiple times in the test where your basically asking for the same thing: "Is this pod running?", it might be easier to just consolidate all into a common function

```
func IsAppRunning(namespace, appName string) (string, error) {
    cmd := "kubectl get pods -n" + namespace + " -o=name -l k8s-app=" + appName + " --field-selector=status.phase=Running --kubeconfig=" + KubeConfigFile
    return RunCommand(cmd)
}
```

Reply:
Fixed